src/NHibernate/Impl/AbstractSessionImpl.cs

```diff
@@ -430,6 +431,7 @@ protected void AfterOperation(bool success)
 				if (!ConnectionManager.IsInActiveTransaction)
 				{
 					ConnectionManager.AfterNonTransactionalQuery(success);
+					ConnectionManager.AfterTransaction();
```

Comment:
Could be moved into `ConnectionManager.AfterNonTransactionalQuery`. This is not the Hibernate way but well, Hibernate does not have that `AfterNonTransactionalQuery`, which in NHibernate just do a log currently.